pkg/oidc/provider/authorize.go

```diff
@@ -0,0 +1,239 @@
+		return
+	}
+
+	// redirect to the redirect_uri with a valid code
+	http.Redirect(w, r, params.redirectURI+"?code="+code+"&state="+params.state, http.StatusFound)
+}
+
+func (h *authorizeHandler) getAndVerifyRancherTokenFromRequest(r *http.Request) (*v3.Token, error) {
+	tokenAuthValue := tokens.GetTokenAuthFromRequest(r)
+	if tokenAuthValue == "" {
+		return nil, fmt.Errorf("rancher token not present")
+	}
+	tokenName, tokenKey := tokens.SplitTokenParts(tokenAuthValue)
+	if tokenName == "" || tokenKey == "" {
+		return nil, fmt.Errorf("can't split rancher token")
+
+	}
+	token, err := h.tokenCache.Get(tokenName)
+	if err != nil {
+		return nil, fmt.Errorf("can't get token: %v", err)
+	}
+	if token.Token != tokenKey {
+		return nil, fmt.Errorf("token doesn't match")
+	}
+	if token.Enabled != nil && !*token.Enabled {
+		return nil, fmt.Errorf("token not enabled")
+	}
+
+	// If the auth provider is specified make sure it exists and enabled.
+	if token.AuthProvider != "" {
+		disabled, err := providers.IsDisabledProvider(token.AuthProvider)
+		if err != nil {
+			return nil, fmt.Errorf("can't check if auth provider is disabled: %v", err)
+		}
+		if disabled {
+			return nil, fmt.Errorf("auth provider is disabled")
+		}
+	}
+
+	authUser, err := h.userLister.Get(token.UserID)
+	if err != nil {
+		return nil, fmt.Errorf("can't get user: %v", err)
+	}
+
+	if authUser.Enabled != nil && !*authUser.Enabled {
+		return nil, fmt.Errorf("user is disabled")
+	}
+
+	return token, nil
+}
+
+// getAuthParamsFromRequest returns the params for the request. OIDC spec says that params can be either in a GET or POST request, so we should check both.
+func getAuthParamsFromRequest(r *http.Request) (*authParams, error) {
+	if r.Method == "POST" {
+		err := r.ParseForm()
+		if err != nil {
+			return nil, err
+		}
+		return &authParams{
+			clientID:            r.Form.Get("client_id"),
+			scopes:              strings.Split(r.Form.Get("scope"), " "),
+			codeChallenge:       r.Form.Get("code_challenge"),
+			codeChallengeMethod: r.Form.Get("code_challenge_method"),
+			nonce:               r.Form.Get("nonce"),
+			state:               r.Form.Get("state"),
+			redirectURI:         r.Form.Get("redirect_uri"),
+			responseType:        r.Form.Get("response_type"),
+		}, nil
+	}
+	if r.Method == "GET" {
+		return &authParams{
+			clientID:            r.URL.Query().Get("client_id"),
+			scopes:              strings.Split(r.URL.Query().Get("scope"), " "),
+			codeChallenge:       r.URL.Query().Get("code_challenge"),
+			codeChallengeMethod: r.URL.Query().Get("code_challenge_method"),
+			nonce:               r.URL.Query().Get("nonce"),
+			state:               r.URL.Query().Get("state"),
+			redirectURI:         r.URL.Query().Get("redirect_uri"),
+			responseType:        r.URL.Query().Get("response_type"),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unsupported method")
```

Comment:
:warning: **untested**

Both `r.Form` and `r.URL.Query()` return an `url.Values`, so this could be rewritten as:


```suggestion
	var values url.Values

	switch r.Method {
	case http.MethodGet:
		values = r.URL.Query()
	case http.MethodPost:
		if err := r.ParseForm(); err != nil {
			return nil, err
		}
		values = r.Form
	default:
		return nil, fmt.Errorf("unsupported method")
	}

	return &authParams{
		clientID:            values.Get("client_id"),
		scopes:              strings.Split(values.Get("scope"), " "),
		codeChallenge:       values.Get("code_challenge"),
		codeChallengeMethod: values.Get("code_challenge_method"),
		nonce:               values.Get("nonce"),
		state:               values.Get("state"),
		redirectURI:         values.Get("redirect_uri"),
		responseType:        values.Get("response_type"),
	}, nil
```